packages/block-editor/src/components/block-visibility/menu-item.js

```diff
@@ -1,96 +1,114 @@
 /**
  * WordPress dependencies
  */
-import { __, sprintf } from '@wordpress/i18n';
+import { __, sprintf, _n } from '@wordpress/i18n';
 import { MenuItem } from '@wordpress/components';
 import { seen, unseen } from '@wordpress/icons';
+import { useState } from '@wordpress/element';
 import { useSelect, useDispatch } from '@wordpress/data';
 import { store as noticesStore } from '@wordpress/notices';
 import { store as keyboardShortcutsStore } from '@wordpress/keyboard-shortcuts';
 
 /**
  * Internal dependencies
  */
-import { cleanEmptyObject } from '../../hooks/utils';
-import { store as blockEditorStore } from '../../store';
+import BlockVisibilityModal from './modal';
+import useBlockVisibility from './use-block-visibility';
+import { VIEWPORT_LABELS } from './utils';
 
 export default function BlockVisibilityMenuItem( { clientIds } ) {
-	const { updateBlockAttributes } = useDispatch( blockEditorStore );
+	const [ isModalOpen, setIsModalOpen ] = useState( false );
 	const { createSuccessNotice } = useDispatch( noticesStore );
-	const blocks = useSelect(
-		( select ) => {
-			return select( blockEditorStore ).getBlocksByClientId( clientIds );
-		},
-		[ clientIds ]
-	);
+
+	const {
+		blocks,
+		isHidden,
+		viewportType,
+		responsiveEditing,
+		toggleVisibility,
+	} = useBlockVisibility( clientIds );
 
 	const listViewShortcut = useSelect( ( select ) => {
 		return select( keyboardShortcutsStore ).getShortcutRepresentation(
 			'core/editor/toggle-list-view'
 		);
 	}, [] );
 
-	const hasHiddenBlock = blocks.some(
-		( block ) => block.attributes.metadata?.blockVisibility === false
-	);
+	const handleToggle = () => {
+		toggleVisibility();
 
-	const toggleBlockVisibility = () => {
-		const isHiding = ! hasHiddenBlock;
-
-		const attributesByClientId = Object.fromEntries(
-			blocks?.map( ( { clientId, attributes } ) => [
-				clientId,
-				{
-					metadata: cleanEmptyObject( {
-						...attributes?.metadata,
-						blockVisibility: isHiding ? false : undefined,
-					} ),
-				},
-			] )
-		);
-		updateBlockAttributes( clientIds, attributesByClientId, {
-			uniqueByBlock: true,
-		} );
+		// Show notice when hiding
+		if ( ! isHidden ) {
+			const viewportLabel = VIEWPORT_LABELS[ viewportType ];
+			const count = blocks.length;
 
-		if ( isHiding ) {
-			if ( blocks.length > 1 ) {
-				createSuccessNotice(
-					sprintf(
-						// translators: %s: The shortcut key to access the List View.
-						__(
-							'Blocks hidden. You can access them via the List View (%s).'
-						),
-						listViewShortcut
+			let message;
+			if ( responsiveEditing ) {
+				message = sprintf(
+					/* translators: 1: Viewport name, 2: List View shortcut. */
+					_n(
+						'Block hidden on %1$s. Access it via List View (%2$s).',
+						'Blocks hidden on %1$s. Access them via List View (%2$s).',
```

Comment:
"Access it" / "Access them" is a bit imperative (especially when translating in my local tongue. Perhaps we can soften it a bit:

```suggestion
						'Block hidden on %1$s. You can access it via List View (%2$s).',
						'Blocks hidden on %1$s. You can access them via List View (%2$s).',
```